File: cpp/rna-transcription/rna_transcription.cpp
#include "rna_transcription.h"

char transcription::to_rna(char dna) {
  switch (dna) {
    case 'G':
      return 'C';
    case 'A':
      return 'U';
    case 'C':
      return 'G';
    case 'T':
      return 'A';
  }
  throw std::invalid_argument("Char must be one of ['G' 'C' 'T' 'A']");
}

std::string transcription::to_rna(std::string dnaStr) {
  std::string rnaStr;
  for (auto dna : dnaStr) {
    rnaStr += to_rna(dna);
  }
  return rnaStr;
}

File: elixir/rna-transcription/lib/rna_transcription.ex
defmodule RnaTranscription do
  @doc """
  Transcribes a single DNA nucleotide to its RNA complement
  """
  @type dna_nucleotide :: ?G | ?C | ?T | ?A
  @type rna_complement :: ?C | ?G | ?A | ?U
  @spec translate_dna_nucleotide(dna_nucleotide()) :: rna_complement()
  def translate_dna_nucleotide(s) do
    case s do
      ?G -> ?C
      ?C -> ?G
      ?T -> ?A
      ?A -> ?U
    end
  end

  @doc """
  Transcribes a character list representing DNA nucleotides to RNA

  ## Examples

  iex> RnaTranscription.to_rna('ACTG')
  'UGAC'
  """
  @spec to_rna([char]) :: [char]
  def to_rna(dna) do
    Enum.map(dna, &RnaTranscription.translate_dna_nucleotide/1)
  end
end
